---
title: "Variant to Gene mapping analysis in the severe asthma UK Biobank GWAS "
author: "Noemi Nicole Piga"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: true
    theme: united
---

```{r, setup, hide = TRUE}
#library reticulate to be able to add chunks in another language
#library(reticulate)
#To run this .rmd file in a terminal being in the project folder:
#export PATH=${PATH}:/cm/shared/apps/R/deps/rstudio/bin/pandoc
#file="./report/Variant_to_Gene_Report.Rmd"
#Rscript -e 'rmarkdown::render("'$file'")'
```

# Rationale
Master Code
<br>
src/Var_to_Gene_pipeline.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/Var_to_Gene_pipeline.sh'), eval=FALSE, hide=TRUE}
```
Prioritise genes using variant-to-gene mapping tools with credible sets variants from fine-mapping results of GWAS on severe asthma in UK Biobank European individuals.
<br>
I used several methods, following the Shrine et al. methodology (https://www.nature.com/articles/s41588-023-01314-0).
<br>
Colocalisation was based on codes from Jing Cheng
Genes with more than X evidence of being mapped with associated variants are selected for following analysis.

# Analysis

## Variant Annotation
<br>
Code
<br>
src/Variant_annotation_FAVOR.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/Variant_annotation_FAVOR.R'), eval=FALSE, hide=TRUE}
```
Variant annotation can be represented in qualitative terms, such as variant effect predictor category or in quantitative
terms with a score including protein function, conservation, epigenetics, integrative function
(ref https://www.nature.com/articles/s41588-020-0676-4).
When dealing with quantitative measures for the same functional category,
these can be summed up using principal components, as implemented for the first time in the STAAR
framework (ref https://www.nature.com/articles/s41588-020-0676-4) and adopted in FAVOR as well.
Annotation Principal Component (aPC) is the first PC among standardised individual score for a specific functional
category. AnnotationPC is transformed in PHRED-scale score, as CADD. CADD authors suggest 15 as a good threshold for functionally
relevant variants (https://cadd.gs.washington.edu/info). Details on thresholds and information used are stored in Supplementary Information (link to the .xlsx file).
<br>
FAVOR aggregates different sources and categories of annotation. I decided to focus on:
 1.Categorical annotation: information from Gencode on coding with exonic details or non-coding variant, gene affected or nearby;

 2.Functional integrative score: integrative scores for different categories;

 3.Clinical annotation: annotation from ClinVar about effect on a disease

## Colocalisation with Quantitative trait loci (QTL) expression analysis
Code
<br>
NB: There are different scripts for this pipeline. Please, refer to 'Var_to_Gene_pipeline.sh' script to have a look at the pipeline workflow.
Codes below are ordered as in their order of appearance in 'Var_to_Gene_pipeline.sh' script.
src/coloc/000_liftover_b38_to_b37_GTExv8.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000_liftover_b38_to_b37_GTExv8.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/000_preprocess_cs.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000_preprocess_cs.R'), eval=FALSE, hide=TRUE}
```
src/coloc/001_submit_GWASpairs.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_submit_GWASpairs.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/001_run_GWASpairs.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_run_GWASpairs.R'), eval=FALSE, hide=TRUE}
```
src/coloc/001_submit_eqtl_lookup_GTEx.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_submit_eqtl_lookup_GTEx.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/001_run_eqtl_lookup_GTEx.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_run_eqtl_lookup_GTEx.R'), eval=FALSE, hide=TRUE}
```
src/coloc/002_prepare_LDinput.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/002_prepare_LDinput.R'), eval=FALSE, hide=TRUE}
```
src/coloc/002_get_LD.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/002_get_LD.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/003_submit_coloc_susie_GTEx.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/003_submit_coloc_susie_GTEx.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/003_run_coloc_susie_GTEx.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/003_run_coloc_susie_GTEx.R'), eval=FALSE, hide=TRUE}
```
src/coloc/004_concat_results.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/004_concat_results.R'), eval=FALSE, hide=TRUE}
```


Quantitative trait loci (QTL) allows us to measure a trait, such as RNA expression of genes or protein levels, in a specific tissue
and test for association of genes within the interested tissue with genetic variant. In other words, the expression trait is
used as a quantitative trait for a genome-wide analysis.
<br>
Looking at QTL and GWAS summary statistics, we can use statistical methods -colocalisation- to assess if both study reveal statistically significant
association in the same loci for the same variant. If so, we can then declare that the variant we found from the GWAS show evidence of association
with a gene in a specific tissue.
<br>
For this reason, QTL colocalisation is used as one of the variant-to-gene mapping analysis.
<br>
Different types of QTL data are available, based on the quantitative trait we study as well as at the type of variant-gene relationship we want to look at.
<br>
To expand on this, expression QTL looks at mRNA expression,; in addition, if the analysis is only for variants near
the gene's transcription start site (usually +/-1Mb), this QTL is called 'cis'-eQTL; if the analysis is for variants further away
that can have an effect on gene expression via the 3D conformation on DNA, they are called 'trans'-eQTL. If, instead, we want
to investigate the role of the variant on the splicing of the mRNA, we call it 's'QTL; they can be both cis or trans- sQTL.
<br>
If we are looking at protein level, we call it 'p'QTL; if looking at methylation quantitative measure, we call it 'm'QTL.
<br>
For my asthma project, I run the following cis-QTL types with the listed dataset:
 1. eQTL: GTExV8, eQTLGen blood, UBC lung eQTL, and U-BIOPRED;

 2. sQTL: GTExV8; pQTL using deCODE plasma;

 3. pQTL: pQTL UKBiobank, SCALLOP pQTL, and INTERVAL pQTL.
<br>
GTExV8 is a collection of eQTL data for significant variant-gene association in 49 tissues with at least 70 samples (https://www.gtexportal.org/home/methods).
<br>
eQTLGen blood is
<br>
UBC lung eQTL is
<br>
U-BIORPED is
<br>
deCODE is
<br>
SCALLOP is
<br>
INTERVAL is
<br>
I used filtered GTExv8 data including variant-gene pairs for each tissue with false discovery rate (FDR) > 0.05, as done by my colleague
Kayesha Coley, who created the filtered files and split into chromosomes. She also added information on allele frequency for European
ancestry individuals.
<br>
For the fine-mapped loci, I created separate files for each locus' credible set (17 locus --> 17 files).

## Polygenic Priority Score (PoPS)


## Rare Variant Analysis


## Nearby Mendelian rare disease-genes


## Nearby Mouse knockout orthologs


# Results

# Conclusion

# Notes