---
title: "Variant to Gene mapping analysis in the severe asthma UK Biobank GWAS "
author: "Noemi Nicole Piga"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: true
    theme: united
---

```{r, setup, hide = TRUE}
#library reticulate to be able to add chunks in another language
#library(reticulate)
#To run this .rmd file in a terminal being in the project folder:
#export PATH=${PATH}:/cm/shared/apps/R/deps/rstudio/bin/pandoc
#file="./report/Cell_tissue_type.Rmd"
#Rscript -e 'rmarkdown::render("'$file'")'
```

# To run this .rmd file in a terminal:
#export PATH=${PATH}:/cm/shared/apps/R/deps/rstudio/bin/pandoc
#file="./report/cell_tissue_type.Rmd"
#module unload R/4.2.1
#module load R/4.1.0
#Rscript -e 'rmarkdown::render("'$file'")'

# Rationale
Prioritise genes using variant-to-gene mapping tools with credible sets variants from fine-mapping results of GWAS on severe asthma in UK Biobank European individuals.
<br>
I used several methods, following the Shrine et al. methodology (https://www.nature.com/articles/s41588-023-01314-0).
<br>
Colocalisation was based on codes from Jing Cheng
Genes with more than X evidence of being mapped with associated variants are selected for following analysis.

# Analysis

## Variant Annotation
Variant annotation can be represented in qualitative terms, such as variant effect predictor category or in quantitative
terms with a score including protein function, conservation, epigenetics, integrative function
(ref https://www.nature.com/articles/s41588-020-0676-4).
When dealing with quantitative measures for the same functional category,
these can be summed up using principal components, as implemented for the first time in the STAAR
framework (ref https://www.nature.com/articles/s41588-020-0676-4) and adopted in FAVOR as well.
Annotation Principal Component (aPC) is the first PC among standardised individual score for a specific functional
category. AnnotationPC is transformed in PHRED-scale score, as CADD. CADD authors suggest 15 as a good threshold for functionally
relevant variants (https://cadd.gs.washington.edu/info). Details on thresholds and information used are stored in Supplementary Information (link to the .xlsx file).
<br>
FAVOR aggregates different sources and categories of annotation. I decided to focus on:
 1.Categorical annotation: information from Gencode on coding with exonic details or non-coding variant, gene affected or nearby;

 2.Functional integrative score: integrative scores for different categories;

 3.Clinical annotation: annotation from ClinVar about effect on a disease

<br>
Code: src/
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Variant_annotation_FAVOR.R'), eval=FALSE, hide=TRUE}
```

## Quantitative trait expression analysis (QTL)
Code: src/
<br>
Structure of the code:
001_submit_eqtl_lookup_GTEx.sh > preprocess_cs.R and 001_run_eqtl_lookup_GTEx.R
<br>
001_submit_GWASpairs.sh > 001_run_GWASpairs.R

```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/src/coloc/001_submit_eqtl_lookup_GTEx.sh'), eval=FALSE, hide=TRUE}
```
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/src/coloc/preprocess_cs.R'), eval=FALSE, hide=TRUE}
```
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/src/coloc/001_run_eqtl_lookup_GTEx.R'), eval=FALSE, hide=TRUE}
```
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/src/coloc/001_submit_GWASpairs.sh'), eval=FALSE, hide=TRUE}
```
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/src/coloc/001_run_GWASpairs.R'), eval=FALSE, hide=TRUE}
```

For this analysis, I run eQTL using GTExV8, eQTLGen blood, and UBC lung eQTL; sQTL using GTExV8; pQTL using deCODE plasma
pQTL, pQTL UKBiobank, SCALLOP pQTL, and INTERVAL pQTL.
<br>
I created separate files for each locus credible set (17 locus --> 17 files).

## Polygenic Priority Score (PoPS)


## Rare Variant Analysis


## Nearby Mendelian rare disease-genes


## Nearby Mouse knockout orthologs


# Results

# Conclusion

# Notes

#comment line