---
title: "Variant to Gene mapping analysis in the severe asthma UK Biobank GWAS "
author: "Noemi Nicole Piga"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: true
    theme: united
---

```{r, setup, hide = TRUE}
#library reticulate to be able to add chunks in another language
#library(reticulate)
#To run this .rmd file in a terminal being in the project folder:
#export PATH=${PATH}:/cm/shared/apps/R/deps/rstudio/bin/pandoc
#file="./report/Variant_to_Gene_Report.Rmd"
#Rscript -e 'rmarkdown::render("'$file'")'
```

# Rationale
Master Code
<br>
src/Var_to_Gene_pipeline.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/Var_to_Gene_pipeline.sh'), eval=FALSE, hide=TRUE}
```
Prioritise genes using variant-to-gene mapping tools with credible sets variants from fine-mapping results of GWAS on severe asthma in UK Biobank European individuals.
<br>
I used several methods, following the Shrine et al. methodology (https://www.nature.com/articles/s41588-023-01314-0).
<br>
Colocalisation was based on codes from Jing Cheng, and as adapted from Alex Williams.
<br>
Genes with more than X evidence of being mapped with associated variants are selected for following analysis.

# Analysis

## Variant Annotation
<br>
Code
<br>
src/Variant_annotation_FAVOR.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/Variant_annotation_FAVOR.R'), eval=FALSE, hide=TRUE}
```
Variant annotation can be represented in qualitative terms, such as variant effect predictor category or in quantitative
terms with a score including protein function, conservation, epigenetics, integrative function
(ref https://www.nature.com/articles/s41588-020-0676-4).
When dealing with quantitative measures for the same functional category,
these can be summed up using principal components, as implemented for the first time in the STAAR
framework (ref https://www.nature.com/articles/s41588-020-0676-4) and adopted in Functional Annotation of Variants Online Resources (FAVOR) as well.
Annotation Principal Component (aPC) is the first PC among standardised individual score for a specific functional
category. AnnotationPC is transformed in PHRED-scale score, as CADD. CADD authors suggest 15 as a good threshold for functionally
relevant variants (https://cadd.gs.washington.edu/info). Details on thresholds and information used are stored in Supplementary Information (link to the .xlsx file).
<br>
FAVOR aggregates different sources and categories of annotation. I decided to focus on:

 1.Categorical annotation: information from Gencode on coding with exonic details or non-coding variant, gene affected or nearby;

 2.Functional integrative score: integrative scores for different categories;

 3.Clinical annotation: annotation from ClinVar about effect on a disease

## Colocalisation with expression quantitative trait loci (QTL) analysis
Code
<br>
NB: There are different scripts for this pipeline. Please, refer to 'Var_to_Gene_pipeline.sh' script to have a look at the pipeline workflow.
Codes below are ordered as in their order of appearance in 'Var_to_Gene_pipeline.sh' script.
src/coloc/000_preprocess_cs.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000_preprocess_cs.R'), eval=FALSE, hide=TRUE}
```
src/coloc/000A_submit_eqtl_gtex_extraction.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000A_submit_eqtl_gtex_extraction.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/000A_eqtl_gtex_extraction.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000A_eqtl_gtex_extraction.R'), eval=FALSE, hide=TRUE}
```
src/coloc/000B_eqtl_gtex_liftover.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000B_eqtl_gtex_liftover.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/000C_submit_eqtl_gtex_conversion.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000C_submit_eqtl_gtex_conversion.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/000C_eqtl_gtex_conversion.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000C_eqtl_gtex_conversion.R'), eval=FALSE, hide=TRUE}
```
src/coloc/001_submit_GWASpairs.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_submit_GWASpairs.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/001_run_GWASpairs.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_run_GWASpairs.R'), eval=FALSE, hide=TRUE}
```
src/coloc/002_prepare_LDinput.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/002_prepare_LDinput.R'), eval=FALSE, hide=TRUE}
```
src/coloc/002_get_LD.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/002_get_LD.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/003_submit_coloc_susie_GTEx.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/003_submit_coloc_susie_GTEx.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/003_run_coloc_susie_GTEx.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/003_run_coloc_susie_GTEx.R'), eval=FALSE, hide=TRUE}
```
src/coloc/000_submit_edit_eQTLGen.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000_submit_edit_eQTLGen.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/000_run_edit_eQTLGen.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000_run_edit_eQTLGen.R'), eval=FALSE, hide=TRUE}
```
src/coloc/001_submit_eqtl_lookup_GTEx.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_submit_eqtl_lookup_GTEx.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/001_run_eqtl_lookup_GTEx.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_run_eqtl_lookup_GTEx.R'), eval=FALSE, hide=TRUE}
```
src/coloc/002_prepare_LDinput_eqtlgen.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/002_prepare_LDinput_eqtlgen.R'), eval=FALSE, hide=TRUE}
```
src/coloc/003_submit_coloc_susie_eQTLGen.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/003_submit_coloc_susie_eQTLGen.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/003_run_coloc_susie_eQTLGen.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/003_run_coloc_susie_eQTLGen.R'), eval=FALSE, hide=TRUE}
```
src/coloc/004_concat_coloc_results.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/004_concat_coloc_results.R'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/000_submit_lookup_lung_eQTL.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/000_submit_lookup_lung_eQTL.sh'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/000_run_lookup_lung_eQTL.r
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/000_run_lookup_lung_eQTL.r'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/002_prepare_LDinput_ubclung.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/002_prepare_LDinput_ubclung.R'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/003_submit_coloc_susie_lung_eQTL.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/003_submit_coloc_susie_lung_eQTL.sh'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/003_run_ coloc_susie_lung_eQTL.r
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/003_run_coloc_susie_lung_eQTL.r'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/004_concat_coloc_susie_ubclung.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/004_concat_coloc_results_ubclung.R'), eval=FALSE, hide=TRUE}
```
Quantitative trait loci (QTL) allows us to measure a trait, such as RNA expression of genes or protein levels, in a specific tissue
and test for association of genes within the interested tissue with genetic variant. In other words, the expression trait is
used as a quantitative trait for a genome-wide analysis.
<br>
Looking at QTL and GWAS summary statistics, we can use statistical methods -colocalisation- to assess if both study reveal statistically significant
association in the same loci for the same variant. If so, we can then declare that the variant we found from the GWAS show evidence of association
with a gene in a specific tissue.
<br>
For this reason, QTL colocalisation is used as one of the variant-to-gene mapping analysis.
<br>
Different types of QTL data are available, based on the quantitative trait we study as well as at the type of variant-gene relationship we want to look at.
<br>
As an example, expression QTL looks at mRNA expression; in addition, if the analysis is only for variants near
the gene's transcription start site (usually +/-1Mb), this QTL is called 'cis'-eQTL; if the analysis is for variants further away
that can have an effect on gene expression via the 3D conformation on DNA, they are called 'trans'-eQTL. If, instead, we want
to investigate the role of the variant on the splicing of the mRNA, we call it 's'QTL; they can be both cis or trans- sQTL.
<br>
If we are looking at protein level, we call it 'p'QTL; if looking at methylation quantitative measure, we call it 'm'QTL.
<br>
For my asthma project, I run the following cis-QTL types with the listed dataset:

 1. cis-eQTL: GTExV8, eQTLGen blood, UBC lung eQTL, and U-BIOPRED;

 3. pQTL: pQTL UKBiobank, SCALLOP pQTL, deCODE plasma

<br>
GTExV8 is a collection of eQTL data for significant variant-gene association in 49 tissues with at least 70 samples (https://www.gtexportal.org/home/methods).
I run co-localisation for the following tissues: 'Artery_Aorta', 'Artery_Coronary', 'Artery_Tibial', 'Colon_Sigmoid', 'Colon_Transverse', 'Esophagus_Gastroesophageal_Junction', 'Esophagus_Muscularis',  'Lung', 'Small_Intestine_Terminal_Ileum', 'Stomach'.
<br>
I used filtered GTExv8 data including variant-gene pairs for each tissue with both either GWAS or eQTL pvalues <= 5x10-6.
Kayesha Coley, my colleague, created the filtered files and split into chromosomes. She also added information on allele frequency for European
ancestry individuals.
<br>
eQTLGen blood is a source for 16,989 eQTL genes from 37 dataset and a total of 31,684 individuals: 11M SNPs with MAF >=1% and within 1Mb from the analysed genes. (https://www.eqtlgen.org/phase1.html)
<br>
UBC lung eQTL is
<br>
TO BE EDITED:
Additional step of p-value correction: since eQTL are less power due ot small sample size,
genome-wide pvalue is not applicable; plus, a lost of genes so there is a larger number of test campared to GWAS, so if using bonferroni the correction would be very small. Instead, we can us eigenMT to corrects for LD strcuture pvalue,
and then usign benjamin hockjerb correction(FDR) to calculate fdr corrected-egineMTadjusted p-value
fdr correction is for accounting multiple testing.
EigenMT and FDR p-value correction were performed by Jing Chen as for the multi-ancestry lung function paper.
Detailed information on the analysis can be found in the main manuscript and related supplementary information (add ref paper).
Jing provided me with the list of significant gene-variant associations for UBC Lung eQTL after p-value correction.
She also provided me backbone scripts to run colocalisation with this eQTL.
<br>
U-BIORPED is (not yet done)

## Colocalisation with protein quantitative trait loci (QTL) analysis
Code
<br>
src/pQTL_coloc/000_preprocess_cs_b38.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/pQTL_coloc/000_submit_lookup_ukbpqtl.sh'), eval=FALSE, hide=TRUE}
```
src/pQTL_coloc/000_submit_lookup_ukbpqtl.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/pQTL_coloc/000_preprocess_cs_b38.R'), eval=FALSE, hide=TRUE}
```
src/pQTL_coloc/001_combine_pqtl.awk
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/pQTL_coloc/001_combine_pqtl.awk'), eval=FALSE, hide=TRUE}
```
deCODE is a free collection of protein plasma pQTL data for 4,907 proteins and 27.2 million tested variants. They used Bonferroni corrected p-value for multiple testing; thus,
0.05/27,200,000=1.8x10−9. They included variants with a MAF > 0.01% and imputation score > 0.9. If the associated variant was within 1Mb from the transcription start site of the gene, the
pQTL was defined as cis, if outside this threshold as trans.
<br>
SCALLOP
<br>
UK Biobank pQTL data

<br>
For the fine-mapped loci, I created separate files for each locus' credible set (17 locus --> 17 files).

## Polygenic Priority Score (PoPS)
Code
src/PoPS/PoPS.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/PoPS/PoPS.sh'), eval=FALSE, hide=TRUE}
```
src/PoPS/submit_pops.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/PoPS/submit_pops.sh'), eval=FALSE, hide=TRUE}
```
src/PoPS/PoPS_summary.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/PoPS/PoPS_summary.R'), eval=FALSE, hide=TRUE}
```
PoPS is a gene prioritisation analysis that uses genome-wide signal from GWAS summary statistics and identified prioritised genes
looking into public bulk and single-cell expression datasets, curated biological pathways, and predicted protein-protein interactions (https://github.com/FinucaneLab/pops; https://www.nature.com/articles/s41588-023-01443-6)
<br>
PoPS is run in three steps, plus a preliminary four step:

 -Step 0: Generate MAGMA scores: run MAGMA to obtain gene association statistics (z-scores) using severe asthma GWAS sumstats and 1000G European individuals;

 -Step 1: Select features: looking at the MAGMA scores, certain features are selected since they are enriched, and stored in the .feature output files;

 -Step 2: Run PoPS: calculate the POP score for each gene using a joint model with a leave-one chromosome out method for the enrichment of all selected features


## Nearby Mendelian rare disease-genes
src/rare_disease/rare_disease.r
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/rare_disease/rare_disease.r'), eval=FALSE, hide=TRUE}
```
Using the Orphadata resource (website link), I analysed if any genes within 500Kb of fine-mapped sentinel variants were associated
with rare diseases showing implication in asthma. In doing so, I filtered the HPO term or any disease including key terms:
'asthma,eosin,immunodef,cili,autoimm,leukopenia,neutropenia'.

## Nearby Mouse knockout orthologs
src/mouse_ko/mouse_ko.r
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/mouse_ko/mouse_ko.r'), eval=FALSE, hide=TRUE}
```
Using the International Mouse Phenotyping (website link), I analysed if any human ortholog gene has been studied in mouse model in the context of respiratory, immune or muscle phenotypes
(top Mouse Phenotype (MP) terms: respiratory system phenotype", "immune system phenotype", "muscle phenotype"). I selected relevant genes within 500Kb from the top causal variant for
each fine mapped locus.

## Rare Variant Analysis

# Results

## Variant Annotation
Looking at the integrative score, I was able to appreciate a positive correlation between aPC.Epigenetics.Active and aPC.Transcription.Factor (0.69);
aPC.Mutation.Density and aPC.Local.Nucleotide.Diversity (0.63); CADD.phred and aPCs.Conservation (0.91). There is also, although weaker, negative correlation
between aPC.Epigenetics.Transcription and aPC.Epigenetics.Repressed (-0.45).
```{r echo=FALSE, out.width = "50%", fig.align = "center"}
knitr::include_graphics("/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/corrplot_integrative_score_15.png")
```
<br>
I obtained a total of 40 genes from the functional annotation analysis. The Venn diagram shows the discovery overlap
between the three different annotation sources:
```{r echo=FALSE, out.width = "30%", fig.align = "center"}
knitr::include_graphics("/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/FAVOR_venn_diagramm_genes.png")
```

# Conclusion

# Notes