---
title: "Variant to Gene mapping analysis in the severe asthma UK Biobank GWAS "
author: "Noemi Nicole Piga"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: true
    theme: united
---

```{r, setup, hide = TRUE}
#library reticulate to be able to add chunks in another language
#library(reticulate)
#To run this .rmd file in a terminal being in the project folder:
#export PATH=${PATH}:/cm/shared/apps/R/deps/rstudio/bin/pandoc
#file="./report/Variant_to_Gene_Report.Rmd"
#Rscript -e 'rmarkdown::render("'$file'")'
```

# Rationale
Identify and prioritise genes using variant-to-gene mapping tools with credible sets variants from fine-mapping
results of GWAS on severe asthma in UK Biobank European individuals.
<br>
Master Code (overall pipeline): src/Var_to_Gene_pipeline.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/Var_to_Gene_pipeline.sh'), eval=FALSE, hide=TRUE}
```
<br>
I used several methods, mainly following the previous study of our group, Shrine et al. 2023 (https://www.nature.com/articles/s41588-023-01314-0).
<br>

# Analysis

## Variant Annotation
<br>
Code
<br>
src/Variant_annotation_FAVOR.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/Variant_annotation_FAVOR.R'), eval=FALSE, hide=TRUE}
```
Variant annotation can be represented in qualitative terms, such as variant effect predictor category or in quantitative
terms with a score including protein function, conservation, epigenetics, integrative function
(ref https://www.nature.com/articles/s41588-020-0676-4).
When dealing with quantitative measures for the same functional category,these can be summed up using principal components,
as implemented for the first time in the STAAR framework (ref https://www.nature.com/articles/s41588-020-0676-4) and adopted in Functional Annotation of Variants Online Resources (FAVOR) as well.
Annotation Principal Component (aPC) is the first PC among standardised individual score for a specific functional
category. AnnotationPC is transformed in PHRED-scale score, as CADD. CADD authors suggest 15 as a good threshold for functionally
relevant variants (https://cadd.gs.washington.edu/info).
<br>
FAVOR aggregates different sources and categories of annotation. I decided to focus on:

 1. Nearest gene for each credible set top PIP variant ("Genecode.Comprehensive.Info");

 2. Functional annotation for all 615 credible set variants looking at either:

    2.1. FANTOM5 Gene Enhancer and/or Gene Promoter ("CAGE.Promoter", "CAGE.Enhancer")

    2.2. Integrative scores > 15 ("CADD.phred","aPC.Protein.Function","aPC.Conservation","aPC.Epigenetics.Active",
    "aPC.Epigenetics.Repressed","aPC.Epigenetics.Transcription","aPC.Local.Nucleotide.Diversity","aPC.Mutation.Density",
    "aPC.Transcription.Factor", "aPC.Mappability");

    2.3. Clinical annotation from ClinVar ("Clinical.Significance", "Gene.Reported")

## Colocalisation with expression quantitative trait loci (QTL) analysis
Code
<br>
src/coloc/000_preprocess_cs.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000_preprocess_cs.R'), eval=FALSE, hide=TRUE}
```
src/coloc/000A_submit_eqtl_gtex_extraction.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000A_submit_eqtl_gtex_extraction.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/000A_eqtl_gtex_extraction.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000A_eqtl_gtex_extraction.R'), eval=FALSE, hide=TRUE}
```
src/coloc/000B_eqtl_gtex_liftover.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000B_eqtl_gtex_liftover.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/000C_submit_eqtl_gtex_conversion.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000C_submit_eqtl_gtex_conversion.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/000C_eqtl_gtex_conversion.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000C_eqtl_gtex_conversion.R'), eval=FALSE, hide=TRUE}
```
src/coloc/001_submit_GWASpairs.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_submit_GWASpairs.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/001_run_GWASpairs.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_run_GWASpairs.R'), eval=FALSE, hide=TRUE}
```
src/coloc/002_prepare_LDinput.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/002_prepare_LDinput.R'), eval=FALSE, hide=TRUE}
```
src/coloc/002_get_LD.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/002_get_LD.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/003_submit_coloc_susie_GTEx.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/003_submit_coloc_susie_GTEx.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/003_run_coloc_susie_GTEx.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/003_run_coloc_susie_GTEx.R'), eval=FALSE, hide=TRUE}
```
src/coloc/000_submit_edit_eQTLGen.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000_submit_edit_eQTLGen.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/000_run_edit_eQTLGen.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/000_run_edit_eQTLGen.R'), eval=FALSE, hide=TRUE}
```
src/coloc/001_submit_eqtl_lookup_GTEx.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_submit_eqtl_lookup_GTEx.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/001_run_eqtl_lookup_GTEx.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/001_run_eqtl_lookup_GTEx.R'), eval=FALSE, hide=TRUE}
```
src/coloc/002_prepare_LDinput_eqtlgen.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/002_prepare_LDinput_eqtlgen.R'), eval=FALSE, hide=TRUE}
```
src/coloc/003_submit_coloc_susie_eQTLGen.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/003_submit_coloc_susie_eQTLGen.sh'), eval=FALSE, hide=TRUE}
```
src/coloc/003_run_coloc_susie_eQTLGen.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/003_run_coloc_susie_eQTLGen.R'), eval=FALSE, hide=TRUE}
```
src/coloc/004_concat_coloc_results.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc/004_concat_coloc_results.R'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/000_submit_lookup_lung_eQTL.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/000_submit_lookup_lung_eQTL.sh'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/000_run_lookup_lung_eQTL.r
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/000_run_lookup_lung_eQTL.r'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/002_prepare_LDinput_ubclung.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/002_prepare_LDinput_ubclung.R'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/003_submit_coloc_susie_lung_eQTL.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/003_submit_coloc_susie_lung_eQTL.sh'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/003_run_ coloc_susie_lung_eQTL.r
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/003_run_coloc_susie_lung_eQTL.r'), eval=FALSE, hide=TRUE}
```
src/coloc_UBClung/004_concat_coloc_susie_ubclung.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/coloc_UBClung/004_concat_coloc_results_ubclung.R'), eval=FALSE, hide=TRUE}
```
src/merge_genes_eqtl_coloc.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/merge_genes_eqtl_coloc.R'), eval=FALSE, hide=TRUE}
```
Quantitative trait loci (QTL) allows us to measure a trait, such as RNA expression of genes or protein levels, in a specific tissue
and test for association of genes within the interested tissue with genetic variant. In other words, the expression trait is
used as a quantitative trait for a genome-wide analysis.
<br>
Looking at QTL and GWAS summary statistics, we can use statistical methods -colocalisation- to assess if both study reveal statistically significant
association in the same loci for the same variant. If so, we can then declare that the variant we found from the GWAS show evidence of association
with a gene in a specific tissue.
<br>
Different types of QTL data are available, based on the quantitative trait we study as well as at the type of
variant-gene relationship we want to look at. As an example, expression QTL looks at mRNA expression; in addition, if the analysis is only for variants near
the gene's transcription start site (usually +/-1Mb), this QTL is called 'cis'-eQTL; if the analysis is for variants further away
that can have an effect on gene expression via the 3D conformation on DNA, they are called 'trans'-eQTL. If, instead, we want
to investigate the role of the variant on the splicing of the mRNA, we call it 's'QTL (cis or trans-sQTL). When analysing protein level,
we identify the measure as it 'p'QTL (methylation quantitative measure, 'm'QTL).
<br>
## eQTL colocalisation
I used cis-QTL data from three datasets: GTExV8, eQTLGen blood, and UBC lung eQTL. I run the analysis if the cis-eQTL showed
significant association within the credible set regions, and I included shared variants present in both eQTL and the
GWAS summary statistic within +/-500Kb from the associated variant. The colocalisation was performed using coloc (https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1004383) or coloc.susie
(http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=PubMed&dopt=Abstract&list_uids=34587156) methods.
<br>
Here, a little description of each cis-eQTL data and the used parameters.
<br>
***GTExV8*** (https://www.gtexportal.org/)
<br>
GTExV8 is a collection of eQTL data for significant variant-gene association in 49 tissues with at least 70 samples (https://www.gtexportal.org/home/methods).
I run co-localisation for the following tissues: 'Artery_Aorta', 'Artery_Coronary', 'Artery_Tibial', 'Colon_Sigmoid',
'Colon_Transverse', 'Esophagus_Gastroesophageal_Junction', 'Esophagus_Muscularis',  'Lung', 'Skin_Not_Sun_Exposed_Suprapubic',
'Skin_Sun_Exposed_Lower_leg','Small_Intestine_Terminal_Ileum', 'Stomach'.
<br>
I used filtered GTExV8 data including variant-gene pairs for each tissue with both GWAS and eQTL pvalues <= 5x10-6.
<br>
***eQTLgen blood eQTLs*** (http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=PubMed&dopt=Abstract&list_uids=34475573)
<br>
eQTLgen blood eQTLs is a source for 16,989 eQTL genes from 37 dataset and a total of 31,684 individuals,
counting 11M SNPs with MAF >=1% and within 1Mb from the analysed genes (https://www.eqtlgen.org/phase1.html).
I used filtered eQTLGen blood variant-gene pairs with both GWAS or eQTL pvalues <= 5x10-6.
<br>
***UBC lung eQTL*** (http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=PubMed&dopt=Abstract&list_uids=23209423)
<br>
UBC Lung eQTL is a collection of lung specimens from 1,111 individuals of European ancestry including 17,178 cis- and 593 trans- lung eQTLs.
<br>
For UBC Lung, raw data were adjusted to find significant p-value: eigenMT to correct for LD structure p-value,
and then benjamini hochberg correction (FDR) to take into account of multiple testing.
<br>
EigenMT and FDR p-value correction were performed by Jing Chen as for the multi-ancestry lung function paper; detailed information
on the analysis can be found in the main manuscript and related supplementary information (https://static-content.springer.com/esm/art%3A10.1038%2Fs41588-023-01314-0/MediaObjects/41588_2023_1314_MOESM1_ESM.pdf).
<br>
Jing provided me with the list of significant gene-variant associations for UBC Lung eQTL after p-value correction.
She also provided me backbone scripts to run colocalisation with this eQTL.
<br>

## pQTL Look-up
Code
<br>
src/pQTL_coloc/000_preprocess_cs_b38.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/pQTL_coloc/000_submit_lookup_ukbpqtl.sh'), eval=FALSE, hide=TRUE}
```
src/pQTL_coloc/000_submit_lookup_ukbpqtl.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/pQTL_coloc/000_preprocess_cs_b38.R'), eval=FALSE, hide=TRUE}
```
src/pQTL_coloc/001_combine_pqtl.awk
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/pQTL_coloc/001_combine_pqtl.awk'), eval=FALSE, hide=TRUE}
```
src/pQTL_coloc/000_submit_lookup_scallop.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/pQTL_coloc/000_submit_lookup_scallop.sh'), eval=FALSE, hide=TRUE}
```
src/pQTL_coloc/000_scallop_lookup.awk
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/pQTL_coloc/000_scallop_lookup.awk'), eval=FALSE, hide=TRUE}
```
src/pQTL_coloc/000_submit_lookup_decode.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/pQTL_coloc/000_submit_lookup_decode.sh'), eval=FALSE, hide=TRUE}
```
src/pQTL_coloc/000_decode_lookup.awk
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/pQTL_coloc/000_decode_lookup.awk'), eval=FALSE, hide=TRUE}
```
I decided to perform a look-up of the 615 credible set variants in three pQTL sources: deCODE plasma, SCALLOP, and UKBiobank pQTL data.
Significant
<br>
***deCODE plasma pQTL*** (https://pubmed.ncbi.nlm.nih.gov/34857953/)
deCODE is a free collection of protein plasma pQTL data for 4,907 proteins and 27.2 million tested variants.
They used Bonferroni corrected p-value for multiple testing (0.05/27,200,000=1.8x10−9). They included variants with a MAF > 0.01% and imputation score > 0.9.
If the associated variant was within 1Mb from the transcription start site of the gene, the pQTL was defined as cis, if outside this threshold as trans.
<br>
***SCALLOP pQTL*** (https://pubmed.ncbi.nlm.nih.gov/33067605/)
<br>
SCALLOP (Systematic and Combined Analysis of Olink Proteins) pQTL combines measures from 90 circulating cardiovascular proteins
in about 30,000 individuals from 13 cohorts.
<br>
***UK Biobank pQTL*** (https://www.nature.com/articles/s41586-023-06592-6)
The Pharma Proteomics Project led to the analysis of 2,923 plasma proteins in 54,219 UK Biobank participants overall.
Within our group, the pQTL association was performed again in 48,195 European samples (as defined in Shrine et al. 2023) and for 1,463 proteins,
including 44.8 million MAC >= 5 variants.
<br>
Variant-protein associations were defined as significant if pQTL p-value was less or equal to 1.8E-9 for deCODE, 5E-8 for SCALLOP,
and 3.41E-5 (0.05/1,463) for UK Biobank pQTL.

## Polygenic Priority Score (PoPS)
Code
src/PoPS/PoPS.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/PoPS/PoPS.sh'), eval=FALSE, hide=TRUE}
```
src/PoPS/submit_pops.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/PoPS/submit_pops.sh'), eval=FALSE, hide=TRUE}
```
src/PoPS/PoPS_summary.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/PoPS/PoPS_summary.R'), eval=FALSE, hide=TRUE}
```
PoPS is a gene prioritisation analysis that uses genome-wide signals from GWAS summary statistics and identified prioritised genes
looking into public bulk and single-cell expression datasets, curated biological pathways, and predicted protein-protein interactions (https://github.com/FinucaneLab/pops; https://www.nature.com/articles/s41588-023-01443-6).
<br>
PoPS includes three steps:

 - Step 0: Generate MAGMA  scores: run MAGMA to obtain gene association statistics (z-scores) using severe asthma GWAS sumstats and 1000G European individuals;

 - Step 1: Select features: looking at the MAGMA scores, certain features are selected since they are enriched, and stored in the .feature output files;

 - Step 2: Run PoPS: calculate the POP score for each gene using a joint model with a leave-one chromosome out method for the enrichment of all selected features

For each variant, I analysed all the genes within 500Kb and reported the top prioritised genes as by PoPS; if no prioritised
genes, I enlarged the genomic window to 1MB.

## Nearby Mendelian rare disease-genes
src/rare_disease/rare_disease.r
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/rare_disease/rare_disease.r'), eval=FALSE, hide=TRUE}
```
Using the ORPHANET resource (https://www.orpha.net/), I analysed if any genes within 500Kb the top causal variant for each fine mapped locus
was associated with rare diseases showing implication in asthma. In doing so, I filtered the Human Phenotype Ontology (HPO)
term or any disease including key terms 'asthma'/'eosin'/'immunodef'/'cili'/'autoimm'/'leukopenia'/'neutropenia'/'macroph'.

## Nearby Mouse knockout orthologs
src/mouse_ko/mouse_ko.r
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/mouse_ko/mouse_ko.r'), eval=FALSE, hide=TRUE}
```
Using the International Mouse Phenotyping (website link), I analysed if any human ortholog gene has been studied in mouse
model in the context of respiratory, immune or muscle phenotypes (top Mouse Phenotype (MP) terms 'respiratory system phenotype'/'immune system phenotype'/'muscle phenotype'.
I selected relevant genes within 500Kb from the top causal variant for each fine mapped locus.

## Rare Variant Analysis
src/rare_variant/000_dataprep_rarevar.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/rare_variant/000_dataprep_rarevar.R'), eval=FALSE, hide=TRUE}
```
src/rare_variant/README.txt
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/rare_variant/README.txt'), eval=FALSE, hide=TRUE}
```
src/rare_variant/submit_rare_variant.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/rare_variant/submit_rare_variant.sh'), eval=FALSE, hide=TRUE}
```
src/rare_variant/create_input_munge_summary_stats.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/rare_variant/create_input_munge_summary_stats.R'), eval=FALSE, hide=TRUE}
```
src/rare_variant/submit_collapsing_bt.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/rare_variant/submit_collapsing_bt.sh'), eval=FALSE, hide=TRUE}
```
I obtained results for single variant and gene-based exome-wide association analysis including 7,413 cases and 36,955 controls, and covariates age_at_recruitment,age2,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10,genetic_sex.
ExWASes were performed using the tool REGENIE v3.1.1 and the Research Analysis Platform (RAP, https://ukbiobank.dnanexus.com/landing).
<br>
To identify rare variant suggestive association and mapped genes, I implemented a look-up analysis for exonic rare variants (MAF < 0.01) within +/- 500Kb from common credible set variants.

## Variant-to-Locus-to-Gene
To combine the results from all the distinct analyses and see which genes were highlighted:
<br>
src/genes_heatmap.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/genes_heatmap.R'), eval=FALSE, hide=TRUE}
```
To understand locus-gene(s) association:
<br>
src/Locus_to_genes_table.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/Locus_to_genes_table.R'), eval=FALSE, hide=TRUE}
```
src/Region_plot_V2G.sh
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/Region_plot_V2G.sh'), eval=FALSE, hide=TRUE}
```
src/Region_plot_V2G.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/Region_plot_V2G.R'), eval=FALSE, hide=TRUE}
```
src/Region_plot_V2G_2.R
```{r, code = readLines('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/Region_plot_V2G_2.R'), eval=FALSE, hide=TRUE}
```


# Results and Conclusion
I identified a total of 98 genes. For each analysis:

 - Nearest gene identified 18 genes;

 - Functional annotation identified 40 genes (FANTOM 14 genes, Integrative scores 37 genes, ClinVar 6 genes);

 - eQTL: GTExV8 8 genes, eQTLgen blood 5 genes, UBC Lung eQTL 3 genes;

 - pQTL: UK Biobank 4 genes, SCALLOP 2 gene (no gene for deCODE);

 - PoPS 16 genes;

 - Mouse knock out 32 genes;

 - Rare disease 6 genes;


No genes were found from the look-up analyses in the exome single or gene-based rare variant analyses.
Two variants (chr6_45913842_C_G,chr7_1489310_G_A) were significant, but not in the sentinel regions; they map to genes TTK and INTS1 respectively.
<br>
Full variant-to-gene results available here:
```{r echo=FALSE}
xfun::embed_file('/home/n/nnp5/PhD/PhD_project/Var_to_Gene/src/report/var2gene_full.xlsx')
```

Out of the 99 genes, 24 genes were supported by at least 2 criteria (Figure 2):

 - 16 genes by two criteria: AC034102.4,AC034114.2,AC044784.1,AC116366.3,CAMK4,HDAC7,HLA-DQA1,HLA-DRB1,HLA-DRB5,IKZF3,IL33,NR1D1,ORMDL3,PGAP3,PPP1R1B,SUOX

 - 6 genes by three criteria: D2HGDH,IL4R,NSMCE1,RORA,RPS26,SMAD3

 - 2 genes by four criteria: BACH2, IL1RL1

```{r, echo=FALSE, out.width = "110%", fig.align = "center"}
knitr::include_graphics("/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/V2G_heatmap_subplots.png")
```
## Locus-to-gene
Below, the region-plots with linkage-disequilibrium measures wth regards to the variant with highest posterior inclusion probability
as by the fine-mapping analysis with gene-mapped in the genomic locus.
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_2_2_102913642_A_AAAAC_102426362_103426362.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_2_2_242692858_C_T_242192858_243192858.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_3_3_33042712_C_T_32547662_33547662.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_5_5_110401872_C_T_109901872_110905675.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_5_5_131885240_C_G_131270805_132496500.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_6_6_32586794_G_T_32086794_33086794.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_6_6_90963614_A_AT_90463614_91463614.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_8_8_81292599_A_C_80792599_81792599.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_9_9_6209697_A_G_5709697_6709697.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_10_10_9064716_C_T_8542744_9564361.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_11_11_76296671_A_G_75796671_76796671.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_12_12_48202941_C_T_47695873_48695873.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_12_12_56435504_C_G_55935504_56935504.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_15_15_61068954_C_T_60569988_61569988.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_15_15_67442596_C_T_66942596_67942596.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_16_16_27359021_A_C_26864806_27864806.pdf){width=100% height=400}
![Alt](/home/n/nnp5/PhD/PhD_project/Var_to_Gene/output/region_plots_V2G/rp_v2g_17_17_38168828_A_G_37573838_38573838.pdf){width=100% height=400}